charts/CSVGenerator.tsx

```diff
@@ -1,91 +1,200 @@
-
-                    if (value != null) {
-                        row.push(value)
-                        rowHasSomeValue = true
-                    } else row.push("")
-                })
-
-                // Only add rows which actually have some data in them
-                if (rowHasSomeValue) rows.push(row.map(csvEscape).join(","))
+
+        return {
+            indexingYears: this.yearIsDayVar!.yearsUniq,
+            titleRow: titleRow
+        }
+    }
+
+    @computed get yearBasedCSVParams(): CSVParams {
+        const titleRow = this.baseTitleRow
+        titleRow.push("Year")
+
+        this.csvDimensions.map(dim =>
+            titleRow.push(csvEscape(dim.fullNameWithUnit))
+        )
+
+        return {
+            indexingYears: uniq(
+                flatten(this.csvDimensions.map(d => d.yearsUniq))
+            ),
+            titleRow: titleRow
+        }
+    }
+
+    @computed get csvParams(): CSVParams {
+        return this.dayIndexedCSV
+            ? this.dayBasedCSVParams
+            : this.yearBasedCSVParams
+    }
+
+    entityCode(entity: string) {
+        return this.chart.entityMetaByKey[entity].code ?? ""
+    }
+
+    private formattedYear(year: number) {
+        return this.chart.formatYearFunction(year)
+    }
+
+    private valueForDimensionEntityYear(
+        dim: ChartDimensionWithOwidVariable,
+        entity: string,
+        year: number
+    ) {
+        return dim.valueByEntityAndYear.get(entity)?.get(year)
+    }
+
+    private yearAndValueForSingleYearDimension(
+        dim: ChartDimensionWithOwidVariable,
+        entity: string
+    ): [string, string | number] | null {
+        let latestYear = undefined
+        let latestValue = undefined
+
+        const latestYearValue = dim.yearAndValueOfLatestValueforEntity(entity)
+
+        if (latestYearValue) {
+            latestYear = dim.formatYear(first(latestYearValue) as number)
+            latestValue = last(latestYearValue)
+        }
+
+        if (latestYear !== undefined && latestValue !== undefined) {
+            return [latestYear, latestValue]
+        } else return null
+    }
+
+    private dimensionsValues(entity: string, year: number) {
+        const values: (string | number)[] = []
+        let rowHasSomeValue = false
+
+        this.allValueDimensions.map(dim => {
+            const value = this.valueForDimensionEntityYear(dim, entity, year)
+
+            if (value !== undefined) {
+                values.push(value)
+                rowHasSomeValue = true
+            } else values.push("")
+        })
+
+        this.singleValueDimensions.map(dim => {
+            const yearAndValue = this.yearAndValueForSingleYearDimension(
+                dim,
+                entity
+            )
+            if (yearAndValue !== null) {
+                values.push(...yearAndValue)
+                rowHasSomeValue = true
+            } else values.push("", "")
+        })
+        return rowHasSomeValue ? values : []
```

Comment:
nit: Would be a little less code and probably wouldn't be slower to ditch `rowHasSomeValue` and return something like `values.some(v => v !== "") ? values : []`

Reply:
Good idea -- removed it.